DOC: Refactor `np.polynomial` docs using `automodule`

rossbar opened the pull request:
Currently, the documentation for each of the `np.polynomial` submodules (e.g. `np.polynomial.chebyshev`, `np.polynomial.laguerre`, etc.) is essentially written in two separate places independently: once in the module doc string, and again in `doc/source/references/routines.polynomial.<poly_type>.rst`.

This approach has the drawback that changes in the code/module docstrings are not automatically reflected in the reference guide documentation. This can be resolved by replacing the duplicate docs in `code/source/references/routines.polynomial.*.rst` with a "dummy" doc that simply uses the sphinx `.. automodule` directive to generate the reference guide documentation from the module doc strings. Doing so has the following advantages:
 * It cuts the size of the documentation for the `np.polynomial` package roughly in half by eliminating redundant `.rst` files
 * It does a better job at guaranteeing that the documentation in the code is in sync with the documentation generated for the reference guide.

In this PR, I have applied the proposed refactor to all six of the submodules in `np.polynomial`. The diffs might be difficult to follow because of all the text deletion. The changes made can be summarized as follows:
 1. The text originally in the `doc/source/reference/routines.polynomials.<poly_type>.rst` has been replaced by an `.. automodule` directive. `.. versionadded` information has been preserved.
 2. The module docstrings for the six polynomial submodules (`chebyshev`, `polynomial`, `hermite`, etc) have been modified to:
   * Replace explicit descriptions of functions with `.. autosummary` directives
   * Reorganized so the convenience classes are the first subsection.
 3. Removed the `routines.polynomials.packages.rst` doc, which essentially set up a toc for the polynomial submodules, and is made redundant by this refactor.

If this is too much all at once, I am happy to close this and create a smaller PR that applies the proposed refactor to only one of the `np.polynomial` submodules to illustrate the main points without such massive diffs.

rossbar committed:
Start at updating np.polynomial docs

rossbar committed:
DOC: Refactor polynomial.chebyshev using automodule.

Reduce duplication in docs by replacing text in
routines.polynomials.<poly_type>.rst with automodule directive.

Modifications in corresponding module docstring to use
autosummary.

rossbar committed:
Refactor hermite_e docs using automodule

rossbar committed:
Applied automodule refactor to remaining polynomial modules.

rossbar committed:
Removed redundant polynomial pkg doc after refactor

rossbar committed:
Added classes and polyutils docs back to toctree

rossbar committed:
Fixed spacing issue in rst role in legendre.py

rossbar committed:
Removed ref to numpy.polynomial module from section title.

The section includes docs for poly1d as well, so title heading
(which is used in the main toc) shouldn't be too specific.

eric-wieser commented on numpy/polynomial/polynomial.py:
```diff
@@ -7,48 +11,63 @@
 
+Classes
+-------
+.. autosummary::
```
My memory of this is that it doesn't actually work, because this rst isn't read until the first pass of autosummary, and there is no recursion so the `polynomial` entry here is never visited.

rossbar replied:
Thanks for the heads up - it's building fine for me with sphinx version 2.4.3 though perhaps that's too new

rossbar replied:
A bit more follow-up: I built the docs with sphinx v2.2.0, which is the minimum required in `conf.py`. The build doesn't complain and the result appears to render properly. I've published the results of running `make html` w/ sphinx 2.2.0 on commit 9bb6a51f9 to [the gh-pages of my fork](https://rossbar.github.io/numpy/reference/routines.polynomials.polynomial.html) so that the results can be more easily compared with the current docs.

mattip commented:
Could you point to an axample of a pre-change page in the [HEAD documentation](http://www.numpy.org/devdocs/) vs a post-change page in the [build artifacts of this PR](https://12026-908607-gh.circle-artifacts.com/0/doc/build/html/index.html) ? I am not sure where to look to get started reviewing this.

rossbar commented:
Sure - the biggest changes to the reference guide are in the `np.polynomial` submodule doc pages. For example, you can compare the [original `np.polynomial.polynomial` doc page](https://numpy.org/devdocs/reference/routines.polynomials.polynomial.html) to [the one generated by this PR](https://rossbar.github.io/numpy/reference/routines.polynomials.polynomial.html). 

Essentially, the content that was originally in `doc/source/reference/routines.polynomials.<polytype>.rst` has been replaced by the module docstring in `numpy/polynomial/<polytype>.py`.

As I mentioned above, if this is too much change all at once, I am happy to revert and confine changes to a single polynomial submodule for demonstration purposes to get the main idea across.

mattip commented:
The two links do not show the same content
- different titles
- different division to categories (the old one matched Poly1D's division) and ordering within them
- different opening paragraph

In general this looks OK to me. The [old toc](https://numpy.org/devdocs/reference/routines.polynomials.html) explicitly calls the links below "modules", so the [new one](https://rossbar.github.io/numpy/reference/routines.polynomials.html) is better.

Here are the links

- [ ] chebyshev [old](https://numpy.org/devdocs/reference/routines.polynomials.chebyshev.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.chebyshev.html) - the new has a note and a reference that do not appear in the old
- [ ] hermite [old](https://numpy.org/devdocs/reference/routines.polynomials.hermite.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.hermite.html) - the new title is missing the "Physicist" part
- [ ] hermite_e [old](https://numpy.org/devdocs/reference/routines.polynomials.hermite_e.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.hermite_e.html) - the new title is strangely formatted and missing the "Probabilists" part
- [ ] laguerre [old](https://numpy.org/devdocs/reference/routines.polynomials.laguerre.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.laguerre.html)
- [ ] legendre [old](https://numpy.org/devdocs/reference/routines.polynomials.legendre.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.legendre.html)
- [ ] package [old](https://numpy.org/devdocs/reference/routines.polynomials.package.html) new was deleted
- [ ] polynomial module [old](https://numpy.org/devdocs/reference/routines.polynomials.polynomial.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.polynomial.html) - now called Power Series

mattip commented:
@charris any thoughts?

rossbar committed:
Update Chebyshev.

rossbar committed:
Updated Hermite.

rossbar committed:
Fixed HermiteE title/formatting

rossbar committed:
Update laguerre

rossbar committed:
Updated polynomial docstring

rossbar commented:
Thanks for taking a look @mattip. I made a few changes to address your comments and make all the docstrings more consistent - see below.

> Here are the links
> 
>     * [ ]  chebyshev [old](https://numpy.org/devdocs/reference/routines.polynomials.chebyshev.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.chebyshev.html) - the new has a note and a reference that do not appear in the old
 - Added back the `.. versionadded`
 - Removed redundant sentence from top of `chebyshev.py` module docstring
> 
>     * [ ]  hermite [old](https://numpy.org/devdocs/reference/routines.polynomials.hermite.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.hermite.html) - the new title is missing the "Physicist" part
 - Re-added "Physicist" part
 - Removed redundant sentence from top of module docstring
> 
>     * [ ]  hermite_e [old](https://numpy.org/devdocs/reference/routines.polynomials.hermite_e.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.hermite_e.html) - the new title is strangely formatted and missing the "Probabilists" part
 - Fixed formatting to match original `HermiteE`
 - Re-added "Probabilists" to title
 - Removed redundant sentence from module docstring
> 
>     * [ ]  laguerre [old](https://numpy.org/devdocs/reference/routines.polynomials.laguerre.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.laguerre.html)
 - Removed redundant sentence from module docstring
> 
>     * [ ]  legendre [old](https://numpy.org/devdocs/reference/routines.polynomials.legendre.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.legendre.html)
> 
>     * [ ]  package [old](https://numpy.org/devdocs/reference/routines.polynomials.package.html) new was deleted
> 
>     * [ ]  polynomial module [old](https://numpy.org/devdocs/reference/routines.polynomials.polynomial.html) and [new](https://rossbar.github.io/numpy/reference/routines.polynomials.polynomial.html) - now called Power 
 - Removed redundant sentence from module docstring

Also, I re-added the `routines.polynomial.package.rst` in order to fix the `numpy.polynomial` reference (for which the link was broken). The "new" `routines.polynomial.package.rst` follows the same pattern as the other `polynomial.*.rst` files - it is just an `automodule` directive linking it to the docstring for the `numpy.polynomial` package. I plan to update the `numpy.polynomial` package docstring as well (see #13113), but will do so in a different PR.

rossbar committed:
Re-added .rst to fix numpy.polynomial link

mattip commented:
Thanks @rossbar
